[1, 2단계 - 체스] 마크(김승현) 미션 제출합니다.

seunghye218 opened the pull request:
안녕하세요 웨지~!
이번 미션은 난이도가 어렵다고 들었고 코드의 양도 많이 필요하다고 느껴졌어요. 마감 시간 내에 동작하는 코드를 구현하는 것이 중요하다 생각하여 구현을 선택하여 집중했어요. 세심한 변수, 메서드 이름 짓기나 캐싱 및 책임 분리 등등 넘어간 부분이 많다고 느꼈지만 구현 후 리팩토링 해야겠다고 생각하였습니다.

마감 기한 때문에 클린 코드를 잘 작성하려면 아래와 같이
1. 프로그래밍 설계 및 구현 역량을 높이거나 
2. 동작하는 코드를 만든 후 리팩토링하는 역량을 높여야 한다.
고 하는데요.

사용하지 않던 언어와 객체 지향 프로그래밍을 배우며 전자, 후자 다 쉽지 않다고 느껴지네요. 그래도 이전 미션들을 할 때보다 조금씩 수월해지고 있다는 것을 위안삼고 위 1, 2 번 역량을 높이기 위한 방법을 모색해 보려고해요 추천하시는 방법이 있으신지 궁금하네요!

seunghye218 committed:
chore: mysql 빌드 의존

- .gitignore docker 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: UserDao 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
docs: 기능 요구 사항 리스트 작성

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 행 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 열 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 좌표 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 기물 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
chore: 패키지 구조 변경

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
docs: 기능 구현 목록 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 기물들을 생성하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
refactor: 기물에서 Point 제거

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 체스판을 생성하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 체스판을 초기화하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 명령을 입력받는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 명령을 찾는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
fix: 킹과 퀸 순서 바뀐 버그 수정

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 기물의 이름 반환하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 체스판을 출력하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 프로그램을 실행하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
docs: 2단계 기능 요구 사항 리스트 작성

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 움직일 기물이 자신 팀의 기물인지 판별하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: Players 생성 및 ChessGame 에 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 비숍이 대각선으로 이동할 수 있는지 판별하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 룩과 퀸이 이동할 수 있는지 판별하는 기능 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 킹과 나이트를 움직일 수 있는지 판별하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 폰을 움직일 수 있는지 판별하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 체스판에서 기물을 원하는 곳으로 이동시킬 수 있다.

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 턴에 맞는 플레이어를 반환하고 턴을 넘기는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 체스 게임에서 기물을 움직이는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
refactor: File 속성 String -> char 변경

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
refactor: 시작 메세지 출력 변경

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 부생성자 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: move 명령 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 빈 체스판을 생성하는 기능 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 나이트 전진 가능 메서드 및 포인트 isAround 메서드 로직 변경

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
refactor: 사용하지 않는 메서드 제거

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 에러 메세지를 출력하는 기능 구현

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
style: 사용하지 않는 임포트 제거

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
fix: 킹 움직임 버그 수정

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
refactor: Board 의 move 메서드 분리

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
refactor: 기물이 null 인 상태를 Empty 객체로 변경

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
feat: 요구 사항에 맞게 프로그램 진행

Co-authored-by: jcoding-play <namejgc@naver.com>

sihyung92 commented on src/test/java/user/UserDaoTest.java:
```diff
@@ -0,0 +1,68 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class UserDaoTest {
```
환경에 따라 실패할 수 도 있는 테스트입니다 (UserDao가 localhost 의존)
테스트가 실패하면 배포 자체가 안 되는 팀이 많습니다. 
임시로 처리해줍시다

```suggestion
@Disabled("localhost 환경에 의존적이므로 비활성화")
class UserDaoTest {
```

seunghye218 replied:
테스트 자바 학습 테스트(?) 를 진행하며 @Disabled 을 쓸데가 있나 싶었는데 이런 경우에 쓸 수 있겠네요!

sihyung92 commented on src/test/java/chess/domain/player/PlayerTest.java:
```diff
@@ -0,0 +1,25 @@
+
+        assertThat(result).isEqualTo(expected);
+    }
+}
```
EndOfLine에 대한 처리가 안 되어있는데, 문제 되는 경우는 거의 없으나 보기 싫으니까 마지막 라인에 개행해줍시다.

seunghye218 replied:
테스트 파일을 만들때 마다 EOL 없이 생성되네요. 이번에 확실하게 ide 설정을 해두었습니다.

seunghye218 replied:
뷰 로직이네요 ㅎㅎ… View 로 옮기도록 하겠습니다.

sihyung92 commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
+docker
```
커밋 단위를 기능 구현 별로 잘 쪼개주셔서 코드 구현 흐름을 보기 편했습니다. 👍

sihyung92 commented on README.md:
```diff
@@ -1,7 +1,119 @@
+- [x] 킹은 주변으로만 이동할 수 있다.
+
+### 나이트
+- [x] 나이트는 한 방향으로 두 칸 이동 후 다른 방향으로 한 칸 이동할 수 있다.
```
뛰어넘을 수 있다는 맥락은 나이트에 고유한 기능이니 추가되면 좋을거 같네요.

seunghye218 replied:
그에 대한 설명이 없어서 추가하면 좋겠네요! 
나이트가 기물을 뛰어넘을 수 있는지 확인하는 부분은 체스판에 존재하는데 체스판에 추가하는건 어떻게 생각하시나요? 현재 기물은 한 점에서 다른 점으로 이동 가능한지 알 수 있는 책임을 지고 있거든요.

체스판
...
  - [x] **[예외 처리]** 기물이 이동할 경로에 기물이 있으면 예외가 발생한다. 
      - [x] 나이트는 다른 기물을 뛰어넘을 수 있다.

sihyung92 commented on src/main/java/chess/domain/piece/Bishop.java:
```diff
@@ -0,0 +1,19 @@
+
+public final class Bishop extends Piece {
+
+    private static final String name = "B";
```
도메인 로직일까요? 😈

seunghye218 replied:
뷰 로직이라 생각이 드네요 😅 View 로 옮기도록 하겠습니다.

sihyung92 commented on src/main/java/chess/domain/player/Player.java:
```diff
@@ -0,0 +1,18 @@
+import chess.domain.piece.Team;
+
+public class Player {
+
```
2줄 개행이 있네요

seunghye218 replied:
종종 놓치는 경우가 있네요. 이번 기회에 두 줄 이상 빈 공백을 처리하는 정규식을 찾았고 자주 포매팅하도록 하겠습니다!

sihyung92 commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+
+    private static final String ERROR_SUFFIX = "[ERROR]";
+
+    // TODO
```
무엇을 Todo 하나요?

seunghye218 replied:
[아래](https://github.com/woowacourse/java-chess/pull/686#discussion_r1533855985) 피드백 해주신 내용과 이중 반복문을 피할 수 있을지 리팩토링하려던 것이었네요. 제출 전에 리팩토링하고 싶었으나 못한 부분들은 따로 적고 지우도록 하겠습니다.

sihyung92 commented on src/main/java/chess/domain/File.java:
```diff
@@ -0,0 +1,57 @@
+
+    private final char file;
+
+    public File(char file) {
+        validateRange(file);
+        this.file = file;
+    }
```
필요한 생성과정과 검증인가요?

A~H로 고정되어 있고 가변적으로 생성될 필요가 없는데요, 불변한 인스턴스 A\~H를 만들어 놓고 재활용하는 방식으로 코드를 구현해보실까요?

seunghye218 replied:
enum으로 변경할 생각이었는데 늦었네요! 다른 Rank, Point, 기물들도 고정된 인스턴스만 사용하기에 캐싱하여 재활용해보려고 합니다. 처음 해보는 패턴이기에 많이 어색한 부분 있으면 고쳐보도록 할게요!

sihyung92 commented on src/main/java/chess/domain/BoardFactory.java:
```diff
@@ -0,0 +1,89 @@
+    public static Map<Point, Piece> createInitialChessBoard() {
+        Map<Point, Piece> board = new HashMap<>();
+
+        board.putAll(createEighthLine());
+        board.putAll(createSeventhLine());
+        IntStream.rangeClosed(3, 6)
+                .boxed()
+                .map(BoardFactory::lineOfEmpty)
+                .forEach(board::putAll);
+        board.putAll(createSecondLine());
+        board.putAll(createFirstLine());
```
8/7줄, 1/2줄 생성이라는 게 가독성이 아주 나쁘진 않은데요,
그래도 백 진영 첫째줄, 둘째줄 생성 / 흑 진영 첫째줄, 둘째줄 생성이라고 하는 게 훨씬 잘 읽힐거 같아요

seunghye218 replied:
그때그때 딱 맞는 함수명이 잘 안떠오르네요 😅

sihyung92 commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,47 @@
+    public void run() {
+        outputView.printGameStart();
+
+        ChessGame game = new ChessGame(BoardFactory.createEmptyBoard());
```
emptyBoard를 할당해두는 절차가 왜 필요한가요?
empty 상태면 start가 아닌 것을 감지하는 로직도 아니고, 실제로 기물을 움직여보면 empty여서 적 진영 기물이라 안 움직인다는 식으로 구성되어있어 유저입장에선 버그로 느껴집니다
<img width="457" alt="image" src="https://github.com/woowacourse/java-chess/assets/51393021/8aaa256d-4247-4c04-ac82-3c98e5d033db">

seunghye218 replied:
[아래 피드백](https://github.com/woowacourse/java-chess/pull/686#discussion_r1533852439)과 같이 유저 입장에서 부드러운 진행이 되도록 하겠습니다.

sihyung92 commented on src/main/java/chess/domain/Command.java:
```diff
@@ -0,0 +1,31 @@
+        return Arrays.stream(values())
+                .filter(command -> command.isExpressionMatch(expression))
+                .findFirst()
+                .orElseThrow();
```
어떤 메시지, 어떤 타입의 예외가 throw 될까요? 이걸로 충분 할까요?

sihyung92 replied:
orElseThrow를 활용하신 부분이 많은데 동일한 리뷰가 모두 적용됩니다

seunghye218 replied:
예외 발생 시 오류를 잘 나타내는 예외와 메시지를 발생하도록 수정하겠습니다.

sihyung92 commented on src/main/java/chess/domain/Command.java:
```diff
@@ -0,0 +1,31 @@
+    MOVE("move"),
+    END("end");
+
+    private final String expression;
```
이 expression이 도메인이 맞나요? 
콘솔 뷰에 특화된 뷰 로직을 도메인이 처리해주고 있지 않나요?
Command는 enum만 가지고 있고 view에서 어떤 enum이 들어왔을지 알아서 결정해서 도메인에 넘겨줘야 재활용성이 높아질거에요.

seunghye218 replied:
뭔가 상태로 빼기에도 애매하다고 생각이 들어 아예 Command 를 지우고 상수와 비교하도록 하였습니다. 이러니 컨트롤러에서는 뷰가 입력을 잘 검증했다 생각하고 로직을 작성하게 되었는데요. 책임 분리가 잘 된것일까요?

sihyung92 commented on src/main/java/chess/domain/Point.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public Point(String point) {
+        this(new File(point.charAt(0)), new Rank(point.charAt(1) - '0'));
```
consoleView에 특화된 로직 아닌가요? 또 charAt 인덱스 접근이라는 위험한 방식을 취하고 있네요 (IndexOutOfExcetpion)

seunghye218 replied:
제출 마감기한 때문에 빨간 약을 선택하고 말았어요... 컨트롤러에서 파싱 후 생성하도록 로직을 바꾸겠습니다.

sihyung92 commented on src/main/java/chess/domain/Point.java:
```diff
@@ -0,0 +1,109 @@
+import chess.domain.piece.Direction;
+import java.util.Objects;
+
+public class Point {
```
Point에 방위와 이동에 대한 역할을 전담시킨건 마음에 드는 설계네요.

sihyung92 commented on src/main/java/chess/domain/Point.java:
```diff
@@ -0,0 +1,109 @@
+
+    public boolean addable(int addFile, int addRank) {
+        return file.addable(addFile) && rank.addable(addRank);
+    }
```
add라는 시그니처는 어색하게 느껴져요. move 아닌가요? board에서는 move라는 개념으로 다루고 있던데요.

seunghye218 replied:
기존 값에 인자로 들어가는 값이 더해져서 그렇게 지었네요. move 로 변경하고 의미가 잘 살도록 파라미터를 Direction 클래스로 변경해보겠습니다.

sihyung92 commented on src/main/java/chess/domain/Board.java:
```diff
@@ -0,0 +1,89 @@
+        Direction route = currentPoint.findRoute(destination);
+        Point nextPoint = currentPoint.add(route.file(), route.rank());
+        while (!nextPoint.equals(destination)) {
+            if (!board.get(nextPoint).equals(new Empty(Team.EMPTY))) {
```
1. new Empty(Team.EMPTY) 는 이 방식으로 밖에 사용 안 될 텐데, 싱글톤으로 만들고 재활용 하는건 어떠세요? Empty.getInstance()
2. NPE를 피하는 습관은 null이 아닌 것이 확실한 걸 앞으로 땡기는 습관입니다. 상수나 null이 아닌 것을 항상 앞에두세요.
```suggestion
            if (!Empty.getInstance().equals(board.get(nextPoint))) {
```

seunghye218 replied:
1. 리팩토링으로 행과 열, 포인트, 각종 기물들을 캐싱해둘 생각이었어요. 기물들의 경우 하나의 체스판에서는 많이 사용되진않지만 리팩토링 하기 쉽다면 확장성을 고려해도 좋다고 생각이 드네요. 다른 기물들도 잘 적용해보겠습니다!
2. 문자열만 의식해왔었네요. 다른 모든 객체에 대해서 NPE를 피하려고 의식해볼게요.

sihyung92 commented on src/main/java/chess/domain/Board.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private void validateMyPiece(Player player, Piece currentPiece) {
+        if (!player.isMyPiece(currentPiece)) {
```
부정연산자를 많이 쓰시는데 가능하다면 부정 연산자의 사용은 지양해주세요. 가독성을 저해합니다

seunghye218 replied:
최대한 지양하도록 하겠습니다.

sihyung92 commented on src/main/java/chess/domain/Board.java:
```diff
@@ -0,0 +1,89 @@
+
+    // TODO 리팩터링
+    private void validatePawn(Point currentPoint, Point destination, Piece currentPiece) {
+        if (currentPiece instanceof Pawn) {
```
리팩토링 대상이시겠지만~~
instanceOf 코드는 가급적 지양해주세요. 
if문도 depth가 깊네요

seunghye218 replied:
넵 책임을 체스판에서 폰으로 옮겨 instanceOf 와 들여쓰기 깊이 문제를 해결하겠습니다.

sihyung92 commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,47 @@
+            try {
+                String readCommand = inputView.readCommand();
+                Command command = Command.from(readCommand);
+                if (command.isEnd()) {
+                    break;
+                }
```
종료시 아무 안내도 없이 종료되네요, 종료 절차가 있는게 좋겠어요

seunghye218 replied:
유저 경험도 신경쓰며 작성하도록 노력하겠습니다.

sihyung92 commented on src/main/java/chess/domain/player/Players.java:
```diff
@@ -0,0 +1,22 @@
+
+    public Players(Player first, Player second) {
+        this.players = List.of(first, second);
+        this.turn = 0;
```
0이 백인지 흑인지 이 코드를 본 사람이 알 수 있나요?
원시값 사용이기도 하고요

seunghye218 replied:
int turn 값을 통해 몇 수 두었는지 알 수 있기 때문에 사용했었는데요. 그러려면 Turn 클래스를 만드는 것이 나아보이네요. 필요하지도 않은 기능이라 Team enum 을 활용하여 알 수 있도록 하겠습니다.

sihyung92 commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+    public void printBoard(Map<Point, Piece> board) {
+        StringBuilder builder = new StringBuilder();
+
+        for (int rank = 8; rank > 0; rank--) {
+            for (char file = 'a'; file <= 'h'; file++) {
```
rank와 file에 대해 view가 정의하고 있네요
이러면 domain 측에서 file, rank를 변경했을 때 view에서는 알게모르게 오류가 발생합니다

seunghye218 replied:
Rank 와 File 에서 변경된다면 오류가 발생하겠네요. 객체에서 값을 가져오도록 변경하겠습니다!

sihyung92 commented on src/test/java/chess/domain/piece/BoardFactoryTest.java:
```diff
@@ -0,0 +1,20 @@
+
+class BoardFactoryTest {
+
+    @Test
+    @DisplayName("기물들을 생성할 수 있다.")
+    void create() {
+        Map<Point, Piece> pieces = BoardFactory.createInitialChessBoard();
+
+        assertThat(pieces).hasSize(64);
+    }
```
Empty까지 기물인가요? 64개는 보드가 잘 생성되었는지 검증하는 테스트네요. 
기물들도 추가해볼까요?

seunghye218 replied:
추가적으로 32개의 빈 기물인지, 2개의 킹이 있는지 검증을 하겠습니다. 모든 기물을 다 확인할 필요까지 있나 고민이 되네요.

sihyung92 replied:
엄밀한 테스트는 시스템의 견고성을 높여주니까 잘 되있으면 당연히 좋겠으나 생산성과 충돌하는 지점이 있습니다 :>
절충점은 스스로 그리고 팀과 함께 찾는거라고 생각합니다

sihyung92 reviewed (changes requested):
안녕하세요 마크~! 리뷰어 웨지입니다.
체스는 정상적으로 잘 동작하네요 👍
TODO 등 스스로 아쉬우실 만한 부분도 많아 보여서, 중간정산으로 request change를 보냅니다.
반영 후 재요청주세요~ (질문은 편하게 해주시고용~)

seunghye218 committed:
style: EOL 추가

Co-authored-by: jcoding-play <namejgc@naver.com>

seunghye218 committed:
test: 아직 사용하지않을 테스트 비활성화

seunghye218 committed:
docs: 나이트 세부 사항 추가

seunghye218 committed:
refactor: Empty 싱글톤으로 변경

- Point findRoute 메서드 리팩토링

seunghye218 committed:
fix: Empty 생성 실패 버그

seunghye218 committed:
refactor: Piece 가 서브 클래스들 생성 책임을 가짐

seunghye218 committed:
refactor: Piece 서브 인스턴스들 캐싱

seunghye218 committed:
refactor: Piece 출력 문자 책임 OutputView 로 이동

seunghye218 committed:
chore: 미사용 import 문 제거

seunghye218 committed:
refactor: File, Rank 클래스 -> enum

seunghye218 committed:
chore: BoardFactoryTest 패키지 변경

- 미사용 import 문 제거

seunghye218 committed:
refactor: 체스판 출력 메서드의 하드코딩 값 제거

seunghye218 committed:
refactor: Point 의 폰 시작 위치 로직 제거

seunghye218 committed:
fix: 체스판 rank1 출력 안되는 문제 및 나이트 움직임 문제 해결

seunghye218 committed:
refactor: 명백한 예외 및 메시지

seunghye218 committed:
refactor: Point 캐싱 및 생성자 팩토리 인자 통일

seunghye218 committed:
test: BoardFactory 체스 초기의 32개의 기물과 32개의 빈자리 생성 테스트

- 나이트 기물 뛰어넘는 동작 테스트 추가

seunghye218 committed:
refactor: point 패키지 생성

- Point, File, Rank 포함

seunghye218 committed:
refactor(BoardFactory): 가독성 위한 함수명 변경 및 빈 라인 생성 로직 분리

seunghye218 committed:
chore: 미사용 import 문 제거

seunghye218 committed:
refactor: move 입력에 따른 Point 생성 로직 분리

seunghye218 committed:
style: 두 줄 이상 공백 삭제

seunghye218 committed:
fix: 체스 빈 라인 잘못 생성하는 문제

seunghye218 committed:
refactor: 컨트롤러 run 메서드 분리

- Command 삭제
- 게임 종료 메시지 추가
- 하드 코딩 값 상수화

seunghye218 committed:
refactor: 매서드 명 isDiagonal -> isSlopeOneDiagonal 으로 명확하게 변경

seunghye218 committed:
refactor(BoardFactory): 빈 보드 생성 접근 제어 변경

seunghye218 committed:
refactor: 폰의 움직임 검증 책임 Board -> Pawn

- instanceof 삭제
- Piece 의 isMovable 메서드 파라미터 하나 추가

seunghye218 committed:
refactor(Players): 턴 타입 int -> Team 변경

- Team 에 opponent 메서드 추가

seunghye218 committed:
test(BoardTest): 폰 시작 위치에서 기물 뛰어넘을 수 없는 테스트 추가

seunghye218 committed:
refactor(Point): add 메소드 시그니처 변경

- add(int addFile, int addRank) -> move(Direction direction)

seunghye218 committed:
refactor: 부정 연산자 제거

seunghye218 committed:
refactor: 기물 움직임 책임 Board -> Player

- 두 Player 가 같은 Board 를 가짐
- Board 게터 방어적 복사
- Players 삭제 및 책임 ChessGame 으로 전가
- 테스트 책임에 맞게 이동

seunghye218 committed:
chore: 주석 삭제

seunghye218 committed:
docs: 리팩토링 후 변경된 사항 업데이트

seunghye218 committed:
refactor: 캐싱된 객체 equals -> == 변경

seunghye218 committed:
refactor: ChessGame 플레이어 move와 턴 넘김 분리

sihyung92 commented on src/main/java/chess/view/PieceCharacters.java:
```diff
@@ -0,0 +1,35 @@
+        return POOL.get(piece);
+    }
+
+    private static Map<Piece, Character> createPool() {
+        final Map<Piece, Character> pieces = new HashMap<>();
+
+        pieces.put(Piece.kingFrom(Team.WHITE), 'k');
+        pieces.put(Piece.kingFrom(Team.BLACK), 'K');
+        pieces.put(Piece.queenFrom(Team.WHITE), 'q');
+        pieces.put(Piece.queenFrom(Team.BLACK), 'Q');
+        pieces.put(Piece.bishopFrom(Team.WHITE), 'b');
+        pieces.put(Piece.bishopFrom(Team.BLACK), 'B');
+        pieces.put(Piece.rookFrom(Team.WHITE), 'r');
+        pieces.put(Piece.rookFrom(Team.BLACK), 'R');
+        pieces.put(Piece.knightFrom(Team.WHITE), 'n');
+        pieces.put(Piece.knightFrom(Team.BLACK), 'N');
+        pieces.put(Piece.pawnFrom(Team.WHITE), 'p');
+        pieces.put(Piece.pawnFrom(Team.BLACK), 'P');
+        pieces.put(Piece.empty(), '.');
+
+        return pieces;
+    }
```
복잡한 static 초기화 작업을 수행해주기 위한 문법으로 static 초기화 블록이 있습니다

```java
    private static final Map<Piece, Character> POOL;

    static {
        final Map<Piece, Character> pieces = new HashMap<>();

        pieces.put(Piece.kingFrom(Team.WHITE), 'k');
        pieces.put(Piece.kingFrom(Team.BLACK), 'K');
        pieces.put(Piece.queenFrom(Team.WHITE), 'q');
        pieces.put(Piece.queenFrom(Team.BLACK), 'Q');
        pieces.put(Piece.bishopFrom(Team.WHITE), 'b');
        pieces.put(Piece.bishopFrom(Team.BLACK), 'B');
        pieces.put(Piece.rookFrom(Team.WHITE), 'r');
        pieces.put(Piece.rookFrom(Team.BLACK), 'R');
        pieces.put(Piece.knightFrom(Team.WHITE), 'n');
        pieces.put(Piece.knightFrom(Team.BLACK), 'N');
        pieces.put(Piece.pawnFrom(Team.WHITE), 'p');
        pieces.put(Piece.pawnFrom(Team.BLACK), 'P');
        pieces.put(Piece.empty(), '.');

        POOL = pieces;
    }
```

내부 동작 상 큰 차이는 없지만 ENUM으로 구현했을수도 있을 거 같아요 ㅎ 
```java
enum PieceCharacters {

    WHITE_KING('k', Piece.kingFrom(Team.WHITE)),
    ...
```

seunghye218 replied:
static 초기화 블록을 다른 크루들의 코드를 보다가 뒤늦게 알아 Point 만 사용했었네요 😅. Enum도 O(1)로 가져오려면 위와 같은 코드를 작성해야 할 것같아 클래스로 작성했어요. 가독성은 열거형이 더 좋네요!

sihyung92 commented on src/main/java/chess/view/PieceCharacters.java:
```diff
@@ -0,0 +1,35 @@
+
+    private static final Map<Piece, Character> POOL = createPool();
+
+    static char characterFrom(Piece piece) {
+        return POOL.get(piece);
+    }
```
Map에 의존하므로 map.get(); 에서 Null을 반환할 수 있음을 감안해야합니당! 
Optional<Char> 라던지, 클라이언트 코드에서 null 처리를 해주던지, 디폴트 값을 반환한다던지 null에 대응할 수 있는 방안을 생각해보셔요.

seunghye218 replied:
널 처리를 자주 인지하고 처리해야겠네요! 다른 로직에서도 비검사 예외를 사용하고 있어 예외를 사용하려고합니다. 이 문제가 생긴다면 게임 진행이 불가능하여 프로그램을 종료하거나 게임을 재시작하는 방식으로 처리해야겠어요.

seunghye218 replied:
사용하는 메서드들이 어떤 예외를 발생시킬 수 있는지 찾아보며 오류를 봤을 때 명백하게 하는 등 좋은 예외처리를 하기 위해 의식적으로 노력해야겠어요.

sihyung92 commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,83 @@
+public class ChessController {
+
+    private static final String COMMAND_END = "end";
+    public static final int DEPARTURE_INDEX = 1;
+    public static final int DESTINATION_INDEX = 2;
```
private으로도 가능하네요. 가급적 제약은 강하게 걸어주고 필요한 경우 풀어주는 것이 좋은 코딩 습관입니다

seunghye218 replied:
상수 추출 단축키가 퍼블릭으로 만들던걸 놓쳤네요... private으로 만들도록 변경하겠씁니다. 점점 인텔리제이 설정이 최적화되고 있는것같네요😀

sihyung92 commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,83 @@
+                    return;
+                }
+                pieceMove(readCommand, game);
+            } catch (RuntimeException e) {
```
39라인은 IllegalArgument로, 여기는 Runtime으로 처리하셨던데 이유가 있나요?
또 RuntimeException으로 처리하면 내가 정의하지 않은 오류 응답이 유저에게 노출될 수 도 있는데, 
내부 구현이 밖에 드러난다는 점에서 보안 취약점으로 분류될 수 있습니다.

IllegalArgumentException이라고 안전한 것은 아니고 가장 확실한 것은 Custom 예외를 사용하는 부분이지만, 
Custom 예외까지 적용하실 필요는 없고 '그렇구나' 인지 정도만 하셔도 좋을거 같아요

seunghye218 replied:
39라인은 try문의 메서드에서 직접 IllegalArgumentException 만 생성하기 때문에 처리했어요. 하지만 Scanner에서 발생할 수 있는 NoSuchElementException, IllegalStateException 을 생각치 못했네요.

일단 던지는 예외가 커스텀 예외인 것 처럼 사용하고 복구 불가능한 오류라면 프로그램을 종료시키고 아니라면 캐치 후 재처리 하는 방식으로 수정해보겠습니다!

sihyung92 commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,83 @@
+            departure = parsePoint(splitCommand[DEPARTURE_INDEX]);
+            destination = parsePoint(splitCommand[DESTINATION_INDEX]);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("잘못된 위치를 입력하였습니다.");
```
이 안에서 발생한 오류가 반드시 잘못된 위치를 입력해서 그랬다고 확신할 수 있나요?

parseInt에서 발생가능한 NumberFormatException, charAt에서 발생가능한 IndexOutOfBoundsException, File.of에서 발생가능한 IllegalArgumentException 세 가지에 대한 처리만 해주어도 충분한거 같아요. (어떤 예외가 발생가능한지는 docs에 정의되어있고, 인텔리제이에서 메서드에 커서를 올린 뒤 f1를 누르면 볼 수 있습니다)

또 에러정보가 풍부했다면 더 좋았겠네요.

```java
        } catch (IllegalArgumentException | IndexOutOfBoundsException e) { // NumberFormatException은 IllegalArgumentExcetpion의 자식 메서드라서 생략됨
            throw new IllegalArgumentException("잘못된 위치를 입력하였습니다. 입력값 : " + readCommand);
        }
```

seunghye218 replied:
사용하는 메서드들이 어떤 예외를 발생시킬 수 있는지 찾아보며 오류를 봤을 때 명백하게 하는 등 좋은 예외처리를 하기 위해 의식적으로 노력해야겠어요.

sihyung92 commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,83 @@
+        game.turnOver();
+    }
+
+    private Point parsePoint(String splitedCommand) {
```
아래가 올바른 표기라고 합니다!
```suggestion
    private Point parsePoint(String splitCommand) {
```

seunghye218 replied:
의미에 맞게 변수명들을 수정할게요!

sihyung92 commented on src/main/java/chess/domain/Board.java:
```diff
@@ -0,0 +1,34 @@
+
+    private final Map<Point, Piece> board;
+
+    public Board(Map<Point, Piece> board) {
+        this.board = board;
+    }
```
Board는 모든 Point에 대한 기물이 있어야 정상적으로 동작하는 코드에요. 임의의 board를 받을 수 있도록 만드는 건 취약한 코드입니다.

public 생성자를 private로 제약하고, 정적 팩토리 메서드 내에서 BoardFactory를 의존하는 건 어떤가요?

seunghye218 replied:
그 부분에 대해 크게 인지하지 않고 작성했던것 같아요. 확실히 문제가 발생할 요지가 많아 보이네요. Board 생성에 대한 책임을 확실하게 팩토리 클래스로 위임하여 모든 Point에 대한 기물을 생성하도록 하는게 오류 발생 가능성을 줄이도록 할게요.

sihyung92 commented on src/main/java/chess/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,67 @@
+    public boolean isMovable(Point currentPoint, Point nextPoint, Piece target) {
+        Team team = getTeam();
+        List<Point> movablePoints = findMovablePoints(currentPoint, team);
+        return movablePoints.contains(nextPoint) &&
+                (movableDiagonal(currentPoint, nextPoint, target) || movableStraight(currentPoint, nextPoint, target));
```
가독성이 떨어지는 조건식입니당.

또 moveDiagonal, movableStraight는 먼저 판단되면 movablePoints를 계산할 필요가 없으므로,
findMovablePoints에 앞서 
`movableDiagonal(currentPoint, nextPoint, target) || movableStraight(currentPoint, nextPoint, target)` 에 대한 early return이 있다면 더욱 좋을 것 같습니다.

seunghye218 replied:
`movableDiagonal` 과 `movableStraight` 은 팀의 진행 방향과 상관없이 진행되기 때문에 `movablePoints.contains(nextPoint)` 의 우선순위가 높아요. 그렇기에 early return 은 힘들것같습니다. 

로직이 이해하기 헷갈리게 작성이 되었다고 생각이 드네요. 메서드를 분리하여 가독성을 높여보겠습니다.

sihyung92 commented on src/main/java/chess/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,67 @@
+
+        if (isStartPosition(currentPoint)) {
+            findMovablePoint(points, currentPoint,
+                    Direction.of(0, TWO_RANK * forwardDirection));
```
해당 코드에선 어떤 Direction으로 가는지 직관적으로 보이는 게 훨씬 좋은데, 굳이 file과 rank를 넘겨 계산하는 방식은 이해가 어렵습니다. 
우테코는 삼항연산자 혐오집단이긴 하지만 어떤 식으로든 방향을 명시적으로 표현해주는게 나아보입니다.
```suggestion
                    forwardDirection ? TWO_UPPER : TWO_LOWER;
```

seunghye218 replied:
해당 코드가 부수 작용도 많고 가독성도 떨어져 리팩토링을 많이 고민했었습니다. 남들과 다른 길을 걸으니 최적화하는데 한계가 느껴지기도 하네요. 최대한 가독성 좋게 리팩토링 해보겠습니다.

sihyung92 commented on src/main/java/chess/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,67 @@
+                    Direction.of(0, TWO_RANK * forwardDirection));
+        }
+        for (Direction direction : Direction.findPawnDirections()) {
+            findMovablePoint(points, currentPoint,
+                    Direction.of(direction.file(), direction.rank() * forwardDirection));
```
이렇게 자동 개행에 의한 개행이 가독성이 떨어지는 경우, 스스로 개행해주면 더 좋습니다
그리고 direction 관련 코드는 위와 동일한 리뷰입니다. 
```suggestion
            findMovablePoint(
               points,
               currentPoint,
               forwardDirection ? direction : direction.reverse() // 애초에 findWhitePawnDirections / findBlackPawnDirections로 갈라 놓는 것도 방법일듯
           );
```

seunghye218 replied:
위의 답변과 같이 전체적으로 리팩토링 해보겠습니다.

sihyung92 commented on src/main/java/chess/domain/player/Player.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private void validateMyPiece(Piece currentPiece) {
+        if (!currentPiece.isSameTeam(team)) {
```
empty도 상대방의 기물이라고 할 수 있나요?

seunghye218 replied:
메시지가 문제를 생각치 못했네요..! 빈 기물도 따로 처리하도록 하겠습니다.

sihyung92 commented on src/main/java/chess/domain/player/Player.java:
```diff
@@ -0,0 +1,65 @@
+        Direction unitDirection = currentPoint.findUnitDirection(destination);
+        Point nextPoint = currentPoint.move(unitDirection);
+        while (nextPoint != destination) {
+            Piece nextPiece = board.get(nextPoint);
+            if (Piece.empty() != nextPiece) {
```
board에서 처리해 줄 수 있겠네요.
```suggestion
            if (board.isEmpty(nextPoint)) {
```

seunghye218 replied:
확실히 보드의 동작으로 볼 수 있겠네요!

sihyung92 commented on src/main/java/chess/domain/point/File.java:
```diff
@@ -0,0 +1,55 @@
+    H('h');
+
+    private static final Map<Character, File> POOL = Map.of(
+            'a', A, 'b', B, 'c', C, 'd', D, 'e', E, 'f', F, 'g', G, 'h', H
```
File의 종류가 늘어날 때 마다 함께 대응해주어야 하는 코드에요.

아래 처럼도 가능해요. (stream을 쓰든 for문을 쓰든간에)
```java
    private static final Map<Character, File> POOL;
    
    static {
      POOL = Arrays.stream(File.values())
          .collect(
              Collectors.toMap(it -> it.file, Function.identity())
          );
    }
```

seunghye218 replied:
이런 방법이 있었네요! Rank 에도 적용해보겠습니다.

sihyung92 commented on src/main/java/chess/domain/point/File.java:
```diff
@@ -0,0 +1,55 @@
+        return addedFile >= minValue() && addedFile <= maxValue();
+    }
+
+    public static char maxValue() {
+        return 'h';
+    }
+
+    public static char minValue() {
+        return 'a';
+    }
```
아래 처럼은 어떠세요?
```suggestion
    public static char maxValue() {
        return H.file;
    }

    public static char minValue() {
        return A.file;
    }
```

seunghye218 replied:
너무 좋네요.

sihyung92 commented on src/main/java/chess/domain/point/Point.java:
```diff
@@ -0,0 +1,108 @@
+public class Point {
+
+    private static final int SlOPE_TWO = 2;
+    private static final EnumMap<File, EnumMap<Rank, Point>> POOL;
+
+    static {
+        POOL = new EnumMap<>(File.class);
+        for (File file : File.values()) {
+            EnumMap<Rank, Point> rankPoints = new EnumMap<>(Rank.class);
+            for (Rank rank : Rank.values()) {
+                rankPoints.put(rank, new Point(file, rank));
+            }
+            POOL.put(file, rankPoints);
+        }
+    }
```
이 캐싱으로 볼 수 있는 성능 이득보다 이 코드를 읽는 개발자가 소모할 시간당 근무수당이 더 높아 보이긴 해요.
굳이 O(1)을 두번 탐색하는 시간보다 new Point(File, Rank)가 더 빠르기도 할거고요, 
메모리상 볼 수 있는 이득은 너무 미미하고요

equals / hashcode를 구현해주시고 ==은 포기하시는 게 더 나은 선택 같습니다. 어차피 다른 개발자들은 Object에서 == 으로 연산할 수 있을거란 기대는 하지 않습니다. (보면 버그인줄 앎)

seunghye218 replied:
굳이 여러 명이 체스 게임을 한다고 가정하는 것이 아니니 그 생각이 들더라고요. 
가독성이 문제라면 file 과 rank 를 문자열 하나로 묶은 키값을 사용하는 것은 어떻게 생각하시나요?

'==' 를 보면 버그로 보는줄 몰랐네요. C 와 많이 달라 헷갈리네요🥲

sihyung92 replied:
리뷰 당시에 야근하러 가야되서 수당얘기를 적은거 같군요 ㅎㅋ
성능을 챙기는 건 좋은 습관이긴 한데요, 현업에서는 이정도 성능은 챙기려고 하지 않는 경우가 많아 드린 리뷰입니다 모든 걸 최적화 하면 좋겠지만 저희의 시간은 곧 비용(인건비)니까요 

Object간의 비교는 .equals로 많이 하는 편이고, equals 대신 실수로 == 비교를 하는 경우 (String에서 특히) 버그가 나는 경험은 흔하다보니 ==을 보면 의심을 하게 되는건 있습니다 ㅎ

seunghye218 replied:
여태 시공간 복잡도만 고려했었는데 이제 생산성있게 안전한 코드를 작성하는게 중요하겠다고 생각이 드네요. 

동일성과 동등성에 대해 추가적으로 학습하였습니다. 포인터 비교와 동등 연산자 오버로딩 등이 떠오르네요.

sihyung92 reviewed (changes requested):
안녕하세요 마크! 리뷰어 웨지입니다.
리뷰 반영 잘 해주셨다는 생각이 들어요 ^_^
한번 더 고고~
반영 후 요청주세요!

seunghye218 committed:
refactor(PieceCharacters): 존재하지 않는 기물 출력 시 게임 종료

- static 초기화 블럭 사용

seunghye218 committed:
refactor(PieceCharacters): 상수 접근 범위 제한

seunghye218 committed:
refactor: 놓친 예외 처리 및 메시지 명확히

seunghye218 committed:
rename: 변수명 수정

seunghye218 committed:
refactor: Board 생성을 확실하게 BoardFactory 로 이동

seunghye218 committed:
refactor: Pawn 의 isMovable 조건문 분리

- Point 의 isAround 매소드 로직 단순화

seunghye218 committed:
refactor: Pawn isMovable 메서드

- Direction 폰 강결합 함수 제거
- Piece getTeam 제거 isWhite 추가

seunghye218 committed:
fix: 존재하지 않는 기물 move 시 잘못된 메시지 출력 문제

seunghye218 committed:
refactor: 체스판의 한 위치가 비어있는지 확인하는 동작 Player -> Board

seunghye218 committed:
refactor: File 및 Rank Pool 생성 로직

seunghye218 committed:
refactor(Point): pool 타입 및 생성 로직 변경

sihyung92 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final Map<Point, Piece> board;
+
+    Board(Map<Point, Piece> board) {
+        this.board = board;
+    }
```
defualt로 처리하셨군요 ㅎ 이것도 방법이지만 의도했던건 아래와 같은 방식이었습니다. 취사선택 하시면 될거 같습니다
```java
    private Board(Map<Point, Piece> board) {
        this.board = board;
    }

     public static Board createInstance(){
          return BoardFactory.createInitialChessBoard()
     } 
```

seunghye218 replied:
피드백을 주셨던 것보다 제가 상대적으로 더 불안전하게 적용했네요 . 모든 기물이 필수적으로 존재해야하니 Board 와 Facotry를 결합하는 것이 defalut 보다 더 안전하겠네요. 

보드와 팩토리 코드 <- 모든 기물 필수적으로 존재해야함
-> 오히려 결합도를 낮추면 문제 가능성 증가  -> 안전하게 코드 작성 위해 클래스 내부 호출

다음 스텝에서 선택하여 적용해보겠습니다.

sihyung92 commented on src/main/java/chess/domain/piece/Queen.java:
```diff
@@ -0,0 +1,25 @@
+        super(team);
+    }
+
+    static Queen from(Team team) {
+        return POOL.get(team);
+    }
```
POOL이 엄청 늘어난거 같은 느낌이 드네요 ㅎ

변경에 취약해서 Piece 생성단의 POOL은 다 걷어내시는게 나을거 같습니다. Team이 하나가 추가되면 몇 개의 코드를 수정해야되는지 상상해보세요. 

java의 GC 메커니즘은 강력해서 공간복잡도를 좋게 하는 게 큰 의미가 있지 않습니다. 대신 버그로 장애 한 번 나면 몇명의 유저라도 떠나가겠죠.

seunghye218 replied:
일반적인 경우 생상성이 중요. 필요치않은 캐싱 -> 생산성 저하
정말로 캐싱이 필요할 때 아래와 같이 확장에도 대비 가능하게 작성 
```
POOL = Arrays.stream(Team.values())
                    .collect(Collectors.toMap(Function.identity(), Queen::new));
```

캐싱을 처음 사용해보고 웨지 덕분에 많은 장단점을 직접 체감하며 기준이 잡혀간다고 생각이 들었어요. 

지속적으로 피드백을 받고 제가 자바 코드를 자바스럽게 작성하고있지 않다고 느껴지네요. 자바라는 언어가 어떤 환경과 목적으로 사용하는지, 지향하는 바가 무엇인지 등 특성과 역사 등을 더 알아봐야겠어요.

sihyung92 commented on src/main/java/chess/domain/point/Point.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public static Point of(File file, Rank rank) {
+        return POOL.get(toKey(file, rank));
```
캐시에 없으면 null을 반환하기보다 새롭게 생성해 반환하고 캐시를 갱신해야 하지 않을까요?

seunghye218 replied:
필드명으로 작성한 것처럼 캐시보다 풀이라고 생각하고 작성한 것을 캐시라고 답변을 하고 있었네요. 코드에서 필요한 모든 file, rank 의 조합으로 포인트를 stream으로 구현했기 때문에 없으면 새로 생성할 부분이 없다고 생각합니다.

나중에 다른 개발자가 수정하고 위 로직이 깨진다고 가정한다면 널을 반환한다는 것을 주석으로 작성하거나 잘못된 file, rank 가 들어온 것이므로 예외를 발생하도록 해야겠네요.

sihyung92 commented on src/main/java/chess/view/PieceCharacters.java:
```diff
@@ -0,0 +1,37 @@
+
+    static char characterFrom(Piece piece) {
+        Character c = POOL.get(piece);
+        if (c == null) {
+            throw new IllegalStateException("체스에 존재하지 않는 기물이 있습니다. 게임을 종료합니다.");
+        }
```
게임을 중단시켜야 할 정도의 예외처리가 필요한 부분은 아니라고 생각해요.

예외 처리는 다양한 방식이 있고(빠른 실패 / 기본값 응답 등등) 모든 상황에 맞는 처리는 없으니 상황에 맞게 써야합니다

seunghye218 replied:
Empty 기물을 그냥 출력하는 방법도 고려해봤지만 오히려 뒤늦게 문제가 발생하고 버그를 찾기 힘들 수 있다고 생각이 들었어요. 말씀하신대로 Board 는 모든 기물이 있어야하고 없다면 게임 로직이 안돌아가므로 그렇게 처리했어요.
그러기 위해서 그전에 널처리를 확실히 했어야했네요...😅

sihyung92 reviewed (approved):
안녕하세요 마크! 리뷰어 웨지입니다~~
코드도 훨씬 가독성이 좋아졌어요 
게임도 잘 돌아갑니다!

작은 리뷰 몇 가지 남겼으니 확인해주시고 다음 스텝에 뵙겠습니다 🙇‍♂️

seunghye218 commented:
전체적으로 제가 안전하고 생산성있게 코드를 작성하지 않았네요. 아마 자바가 지향하는 방향으로 작성하지않은 것과 같은 것 같아요. 이후 코드 작성시 그런 부족한 부분들을 더 고려하여 작성해야겠습니다🧐.
